File: ModularGeometrySynth/Sources/ModularGeometrySynth/Model/ConnectionRules.swift
import Foundation

// Pure function: can these two ports be connected?
func canConnect(fromSignal: SignalType, toSignal: SignalType) -> Bool {
    switch (fromSignal, toSignal) {
    case (.audio,    .audio):   return true
    case (.control,  .control): return true
    case (.trigger,  .trigger): return true
    case (.geometry, .geometry): return true
    case (.geometry, .control): return true  // geometry → control is allowed (e.g. radius → frequency)
    default: return false
    }
}

func canConnect(from: PortDefinition, to: PortDefinition) -> Bool {
    guard from.direction == .output, to.direction == .input else { return false }
    return canConnect(fromSignal: from.signalType, toSignal: to.signalType)
}

// Find the output port def and input port def for a potential edge
func resolveConnection(
    fromNodeId: String, fromPortId: String,
    toNodeId: String,   toPortId: String,
    project: ModularProject
) -> Bool {
    guard
        let fromNode = project.nodes.first(where: { $0.id == fromNodeId }),
        let toNode   = project.nodes.first(where: { $0.id == toNodeId }),
        let fromDef  = nodeRegistry[fromNode.type],
        let toDef    = nodeRegistry[toNode.type],
        let fromPort = fromDef.outputs.first(where: { $0.id == fromPortId }),
        let toPort   = toDef.inputs.first(where:  { $0.id == toPortId })
    else { return false }
    return canConnect(from: fromPort, to: toPort)
}

File: src/patch/connectionRules.ts
import type { SignalType, ModularProject } from './types'
import { nodeRegistry } from './nodeRegistry'

export function canConnectSignals(from: SignalType, to: SignalType): boolean {
  if (from === to) return true
  if (from === 'geometry' && to === 'control') return true // geometry → control allowed
  return false
}

export function canConnect(
  fromNodeId: string, fromPortId: string,
  toNodeId: string,   toPortId: string,
  project: ModularProject,
): boolean {
  const fromNode = project.nodes.find(n => n.id === fromNodeId)
  const toNode   = project.nodes.find(n => n.id === toNodeId)
  if (!fromNode || !toNode) return false
  const fromDef = nodeRegistry[fromNode.type]
  const toDef   = nodeRegistry[toNode.type]
  if (!fromDef || !toDef) return false
  const fromPort = fromDef.outputs.find(p => p.id === fromPortId)
  const toPort   = toDef.inputs.find(p => p.id === toPortId)
  if (!fromPort || !toPort) return false
  return canConnectSignals(fromPort.signalType, toPort.signalType)
}

export function mapRange(
  value: number,
  inMin: number, inMax: number,
  outMin: number, outMax: number,
): number {
  if (inMax === inMin) return outMin
  const t = (value - inMin) / (inMax - inMin)
  return outMin + Math.max(0, Math.min(1, t)) * (outMax - outMin)
}
